File: guests/go/dnd/dnd.go
// The drag-and-drop scene (tools/scenes/dnd.steps; docs/dnd-plan.md D1, D8).
// THE ROOT IS A ROW so column#0 is the reorderable For's container.
package dnd

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"

	kaya "dev.kaya/bindings/go"
)

//go:generate go run dev.kaya/cmd/kaya-gen -type Item -key string
type Item struct {
	Title string
}

const noteID = "dev.kaya/note"

func word(op kaya.Op) string {
	switch op {
	case kaya.OpCopy:
		return "copy"
	case kaya.OpMove:
		return "move"
	default:
		return "none"
	}
}

func keyWord(keys []any) string {
	if len(keys) == 0 {
		return ""
	}
	return fmt.Sprint(keys[0])
}

// Where the harness's $TMP points on each platform, the picker guest's own
// shape (tools/check-go-env.py holds it).
func sceneRoot() string {
	switch runtime.GOOS {
	case "android":
		root := kaya.Env("EXTERNAL_STORAGE")
		if root == "" {
			root = "/sdcard"
		}
		return filepath.Join(root, "Documents")
	case "ios":
		return filepath.Join(kaya.Env("HOME"), "Documents")
	}
	return os.TempDir()
}

// The file the scene drops as a FOREIGN source (D6), written by the guest
// at $TMP/kaya-dnd-$PID/dropped.txt — the picker and clipboard scenes'
// convention.
func writeDroppedFile() {
	dir := filepath.Join(sceneRoot(), fmt.Sprintf("kaya-dnd-%d", os.Getpid()))
	if err := os.MkdirAll(dir, 0o755); err != nil {
		panic(err)
	}
	if err := os.WriteFile(filepath.Join(dir, "dropped.txt"), []byte("dropped bytes"), 0o644); err != nil {
		panic(err)
	}
}

func readBack(f kaya.PickedFile) string {
	handle, _, err := f.Open(kaya.FileModeRead)
	if err != nil {
		return fmt.Sprintf("open failed: %v", err)
	}
	defer handle.Close()
	body, err := io.ReadAll(handle)
	if err != nil {
		return fmt.Sprintf("read failed: %v", err)
	}
	return string(body)
}

func App() *kaya.App {
	writeDroppedFile()
	app := kaya.NewApp()

	app.Build(func(tx *kaya.Tx) {
		items := ItemCollection(tx)
		items2 := ItemCollection(tx)
		dropStatus := tx.Signal("no drop yet")
		dragStatus := tx.Signal("no drag yet")
		sourceText := tx.Signal("hello")
		textTarget := tx.Signal("text target")
		noteTarget := tx.Signal("note target")
		filesTarget := tx.Signal("files target")

		var source, textID, noteID2, filesID kaya.Widget
		var rows *itemRowsFor
		var rowLabel, itemLabel kaya.Node
		tx.Window(0).Title("dnd")
		tx.Mount(tx.Row(func() {
			rows = ItemRows(tx, items)
			for row := range rows.All() {
				rowLabel = row.Label(row.Title())
				row.SetA11yID(rowLabel, "row")
			}
			tx.SetA11yID(rows.Widget(), "rows")
			tx.Column(func() {
				source = tx.Label(sourceText) // label#0
				textID = tx.Label(textTarget).
					Accepts(kaya.AcceptText).
					DropTarget(kaya.OpCopy) // label#1
				noteID2 = tx.Label(noteTarget).
					Accepts(noteID).
					DropTarget(kaya.OpCopy, kaya.OpMove) // label#2
				filesID = tx.Label(filesTarget).
					Accepts(kaya.AcceptFiles).
					DropTarget(kaya.OpCopy) // label#3
				tx.Label(dropStatus) // label#4
				tx.Label(dragStatus) // label#5
			})
			// THE TEMPLATE ZONE (docs/dnd-plan.md §4): every stamped item
			// is a text destination, and its payload IS the row's own
			// field — resolved per copy, re-declared when it changes.
			itemRows := ItemRows(tx, items2)
			for row := range itemRows.All() {
				itemLabel = row.Label(row.Title())
				row.SetA11yID(itemLabel, "item")
				row.SetAccepts(itemLabel, kaya.AcceptText)
				row.SetDropTarget(itemLabel, kaya.OpCopy)
				row.Draggable(itemLabel).Text(row.Title()).Allow(kaya.OpCopy).Declare()
			}
			tx.SetA11yID(itemRows.Widget(), "items")
			tx.Button("rename y", func(tx *kaya.Tx) { // button#0
				items2.Update(tx, "y", Item{Title: "yy"})
			})
		}))
		tx.Draggable(source).
			Text("hello").
			Custom(noteID, []byte("note!")).
			Allow(kaya.OpCopy).
			Allow(kaya.OpMove).
			Declare()
		rows.Reorderable(true)

		dropped := func(name string, target kaya.Signal[string]) func(*kaya.Tx, kaya.Dropped) {
			return func(tx *kaya.Tx, d kaya.Dropped) {
				op := word(d.Operation)
				switch clip := d.Clip.(type) {
				case kaya.TextClip:
					tx.Write(dropStatus, fmt.Sprintf("%s got text %s (%s)", name, clip.Text, op))
					tx.Write(target, clip.Text)
				case kaya.CustomClip:
					tx.Write(dropStatus, fmt.Sprintf("%s got %s %d bytes (%s)", name, clip.ID, len(clip.Bytes), op))
				case kaya.FilesClip:
					// A dropped file IS a picked file (D6): read it back
					// through the same table the picker fills.
					said := ""
					for i, f := range clip.Files {
						if i > 0 {
							said += ", "
						}
						said += fmt.Sprintf("%s %s", f.Name, readBack(f))
					}
					tx.Write(dropStatus, fmt.Sprintf("%s got %s (%s)", name, said, op))
				default:
					tx.Write(dropStatus, fmt.Sprintf("%s got other (%s)", name, op))
				}
				// A same-app MOVE removes its original in the same batch (D2).
				if d.Operation == kaya.OpMove {
					tx.Write(sourceText, "moved out")
					tx.Draggable(source).Declare()
				}
			}
		}
		app.OnDrop(textID, dropped("text target", textTarget))
		app.OnDrop(noteID2, dropped("note target", noteTarget))
		app.OnDrop(filesID, dropped("files target", filesTarget))
		app.OnDragEnded(source, func(tx *kaya.Tx, op kaya.Op) {
			tx.Write(dragStatus, fmt.Sprintf("drag ended %s", word(op)))
		})
		app.OnDropNode(itemLabel, func(tx *kaya.Tx, keys []any, d kaya.Dropped) {
			op := word(d.Operation)
			if clip, ok := d.Clip.(kaya.TextClip); ok {
				tx.Write(dropStatus, fmt.Sprintf("item %s got text %s (%s)", keyWord(keys), clip.Text, op))
				return
			}
			tx.Write(dropStatus, fmt.Sprintf("item %s got other (%s)", keyWord(keys), op))
		})
		nodeEnded := func(what string) func(*kaya.Tx, []any, kaya.Op) {
			return func(tx *kaya.Tx, keys []any, op kaya.Op) {
				tx.Write(dragStatus, fmt.Sprintf("%s %s drag ended %s", what, keyWord(keys), word(op)))
			}
		}
		app.OnDragEndedNode(itemLabel, nodeEnded("item"))
		app.OnDragEndedNode(rowLabel, nodeEnded("row"))
		// The moved row's key rides as the kaya-private custom
		// representation; the anchor is the row it landed on (D8).
		rows.OnDrop(func(tx *kaya.Tx, d kaya.Dropped) {
			clip, ok := d.Clip.(kaya.CustomClip)
			if !ok || len(d.Anchor) == 0 {
				return
			}
			anchor, ok := d.Anchor[0].(string)
			if !ok {
				return
			}
			if d.Before {
				items.MoveBefore(tx, string(clip.Bytes), anchor)
			} else {
				items.MoveAfter(tx, string(clip.Bytes), anchor)
			}
		})

		for _, key := range []string{"a", "b", "c"} {
			items.Insert(tx, key, Item{Title: key})
		}
		for _, key := range []string{"x", "y"} {
			items2.Insert(tx, key, Item{Title: key})
		}
	})

	return app
}

File: guests/java/dev/kaya/guests/Dnd.java
package dev.kaya.guests;

import dev.kaya.KayaApp;
import dev.kaya.KayaGen;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiConsumer;

/**
 * The drag-and-drop scene from the JVM — guests/rust/dnd.rs,
 * tools/scenes/dnd.steps. THE ROOT IS A ROW so column#0 is the
 * reorderable For's container.
 */
public final class Dnd {
    @KayaGen(key = "String")
    record DndItem(String title) {}

    /** Java lambdas cannot assign captured locals. */
    /** The TRACE's own slots: a For's body runs once where the compiler
     * cannot see it, and the template zone is outside the ruling. */
    private static final class Refs {
        KayaApp.Node rowNode, itemNode;
    }

    /** What the row body declared, on its way out to the handlers. */
    private record Targets(KayaApp.Widget source, KayaApp.Widget text,
            KayaApp.Widget note, KayaApp.Widget files, KayaApp.Widget rows) {}

    private static final String NOTE_ID = "dev.kaya/note";

    private static String word(KayaApp.Op op) {
        if (op == KayaApp.Op.COPY) {
            return "copy";
        }
        if (op == KayaApp.Op.MOVE) {
            return "move";
        }
        return "none";
    }

    private static String keyWord(List<Object> keys) {
        return keys.isEmpty() ? "" : String.valueOf(keys.get(0));
    }

    // The file the scene drops as a FOREIGN source (D6), written by the
    // guest at $TMP/kaya-dnd-$PID/dropped.txt — the picker and clipboard
    // scenes' convention, and the same place the interpreters expand $TMP
    // to. TMPDIR FIRST: java.io.tmpdir ignores it on macOS
    // (docs/traps.md, "java.io.tmpdir").
    private static void writeDroppedFile() {
        String tmp = System.getenv("TMPDIR");
        if (tmp == null || tmp.isEmpty()) {
            tmp = System.getProperty("java.io.tmpdir");
        }
        if (System.getProperty("java.specification.vendor", "")
                .contains("Android")) {
            String ext = System.getenv("EXTERNAL_STORAGE");
            tmp = (ext == null || ext.isEmpty() ? "/sdcard" : ext) + "/Documents";
        }
        try {
            Path dir = Paths.get(tmp, "kaya-dnd-" + pid());
            Files.createDirectories(dir);
            Files.write(dir.resolve("dropped.txt"),
                    "dropped bytes".getBytes(StandardCharsets.UTF_8));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    /** This process's id, without NAMING {@code ProcessHandle}: the Android
     * SDK has no such class, so naming it is a compile error there. */
    private static long pid() {
        try {
            return Long.parseLong(
                    new java.io.File("/proc/self").getCanonicalFile().getName());
        } catch (Exception noProc) {
            // macOS and Windows have no /proc; ask the JDK.
            try {
                Class<?> handle = Class.forName("java.lang.ProcessHandle");
                Object current = handle.getMethod("current").invoke(null);
                return (Long) handle.getMethod("pid").invoke(current);
            } catch (Exception e) {
                throw new IllegalStateException(
                        "kaya: cannot determine this process's id", e);
            }
        }
    }

    private static String readBack(KayaApp.PickedFile file) {
        try (InputStream in = file.open(KayaApp.FileMode.READ).stream()) {
            return new String(in.readAllBytes(), StandardCharsets.UTF_8);
        } catch (IOException e) {
            return "open failed: " + e.getMessage();
        }
    }

    public static void app() {
        writeDroppedFile();
        KayaApp app = new KayaApp();

        app.build(tx -> {
            var items = DndItemKaya.collection(tx);
            var items2 = DndItemKaya.collection(tx);
            var dropStatus = tx.signal("no drop yet");
            var dragStatus = tx.signal("no drag yet");
            var sourceText = tx.signal("hello");
            var textTarget = tx.signal("text target");
            var noteTarget = tx.signal("note target");
            var filesTarget = tx.signal("files target");

            Refs refs = new Refs();
            tx.window(0).title("dnd");
            var built = tx.row(cells -> {
                // The For opens when rows() is called, so it is minted
                // INSIDE the row: a window prop emitted between the open
                // and the template's end is refused by the core.
                var rows = DndItemKaya.rows(tx, items);
                for (var row : rows) {
                    refs.rowNode = row.label(row.title);
                    row.setA11yId(refs.rowNode, "row");
                }
                KayaApp.Widget rowsWidget = rows.handle;
                tx.setA11yId(rowsWidget, "rows");
                var targets = tx.column(col -> {
                    KayaApp.Widget sourceLabel = tx.label(sourceText); // label#0
                    KayaApp.Widget textTargetLabel = tx.label(textTarget); // label#1
                    tx.setAccepts(textTargetLabel, KayaApp.ACCEPT_TEXT);
                    tx.setDropTarget(textTargetLabel, KayaApp.Op.COPY);
                    KayaApp.Widget noteTargetLabel = tx.label(noteTarget); // label#2
                    tx.setAccepts(noteTargetLabel, NOTE_ID);
                    tx.setDropTarget(noteTargetLabel, KayaApp.Op.COPY, KayaApp.Op.MOVE);
                    KayaApp.Widget filesTargetLabel = tx.label(filesTarget); // label#3
                    tx.setAccepts(filesTargetLabel, KayaApp.ACCEPT_FILES);
                    tx.setDropTarget(filesTargetLabel, KayaApp.Op.COPY);
                    tx.label(dropStatus); // label#4
                    tx.label(dragStatus); // label#5
                    return new Targets(sourceLabel, textTargetLabel, noteTargetLabel,
                            filesTargetLabel, rowsWidget);
                }).value();
                // THE TEMPLATE ZONE (docs/dnd-plan.md §4): every stamped
                // item is a text destination, and its payload IS the
                // row's own field — resolved per copy, re-declared when
                // the field changes — column#2.
                var itemRows = DndItemKaya.rows(tx, items2);
                for (var row : itemRows) {
                    refs.itemNode = row.label(row.title);
                    row.setA11yId(refs.itemNode, "item");
                    row.setAccepts(refs.itemNode, KayaApp.ACCEPT_TEXT);
                    row.setDropTarget(refs.itemNode, KayaApp.Op.COPY);
                    row.draggable(refs.itemNode).text(row.title).allow(KayaApp.Op.COPY).declare();
                }
                tx.setA11yId(itemRows.handle, "items");
                tx.button("rename y", t -> // button#0
                        items2.update(t, "y", new DndItem("yy")));
                return targets;
            });
            tx.mount(built.id());
            Targets targets = built.value();
            KayaApp.Widget source = targets.source();
            tx.draggable(source)
                    .text("hello")
                    .custom(NOTE_ID, "note!".getBytes(StandardCharsets.UTF_8))
                    .allow(KayaApp.Op.COPY)
                    .allow(KayaApp.Op.MOVE)
                    .declare();
            tx.setReorderable(targets.rows(), true);

            app.onDrop(targets.text(), dropped(app, "text target", textTarget, dropStatus,
                    sourceText, source));
            app.onDrop(targets.note(), dropped(app, "note target", noteTarget, dropStatus,
                    sourceText, source));
            app.onDrop(targets.files(), dropped(app, "files target", filesTarget, dropStatus,
                    sourceText, source));
            app.onDragEnded(source, (t, op) ->
                    t.write(dragStatus, "drag ended " + word(op)));
            app.onDrop(refs.itemNode, (t, keys, d) -> {
                String op = word(d.operation());
                if (d.clip() instanceof KayaApp.Representation.Text text) {
                    t.write(dropStatus, "item " + keyWord(keys) + " got text "
                            + text.value() + " (" + op + ")");
                } else {
                    t.write(dropStatus, "item " + keyWord(keys) + " got other ("
                            + op + ")");
                }
            });
            app.onDragEnded(refs.itemNode, nodeEnded("item", dragStatus));
            app.onDragEnded(refs.rowNode, nodeEnded("row", dragStatus));
            // The moved row's key rides as the kaya-private custom
            // representation; the anchor is the row it landed on (D8).
            app.onDrop(targets.rows(), (t, d) -> {
                if (!(d.clip() instanceof KayaApp.Representation.Custom moved)
                        || d.anchor().isEmpty()
                        || !(d.anchor().get(0) instanceof String anchor)) {
                    return;
                }
                String key = new String(moved.bytes(), StandardCharsets.UTF_8);
                if (d.before()) {
                    items.moveBefore(t, key, anchor);
                } else {
                    items.moveAfter(t, key, anchor);
                }
            });

            for (String key : new String[] {"a", "b", "c"}) {
                items.insert(tx, key, new DndItem(key));
            }
            for (String key : new String[] {"x", "y"}) {
                items2.insert(tx, key, new DndItem(key));
            }
            return null;
        });

        app.dispatchLoop();
    }

    private static KayaApp.DragEndedHandler nodeEnded(
            String what, KayaApp.Signal<String> dragStatus) {
        return (t, keys, op) -> t.write(dragStatus,
                what + " " + keyWord(keys) + " drag ended " + word(op));
    }

    private static BiConsumer<KayaApp.Tx, KayaApp.Dropped> dropped(
            KayaApp app, String name, KayaApp.Signal<String> target,
            KayaApp.Signal<String> dropStatus, KayaApp.Signal<String> sourceText,
            KayaApp.Widget source) {
        return (t, d) -> {
            String op = word(d.operation());
            if (d.clip() instanceof KayaApp.Representation.Text text) {
                t.write(dropStatus, name + " got text " + text.value() + " (" + op + ")");
                t.write(target, text.value());
            } else if (d.clip() instanceof KayaApp.Representation.Custom custom) {
                t.write(dropStatus, name + " got " + custom.id() + " "
                        + custom.bytes().length + " bytes (" + op + ")");
            } else if (d.clip() instanceof KayaApp.Representation.Files files) {
                // A dropped file IS a picked file (D6): read it back
                // through the same table the picker fills.
                List<String> said = new ArrayList<>();
                for (KayaApp.PickedFile f : files.value()) {
                    said.add(f.name() + " " + readBack(f));
                }
                t.write(dropStatus, name + " got " + String.join(", ", said)
                        + " (" + op + ")");
            } else {
                t.write(dropStatus, name + " got other (" + op + ")");
            }
            // A same-app MOVE removes its original in the same batch (D2).
            if (d.operation() == KayaApp.Op.MOVE) {
                t.write(sourceText, "moved out");
                t.draggable(source).declare();
            }
        };
    }

    private Dnd() {}
}
